tough/src/editor/signed.rs

```diff
@@ -219,22 +209,43 @@ impl SignedRepository {
                 continue;
             };
 
-            // If the entry is a file, get the filename
-            let file_name = entry
+            // Call the requested function to manipulate the path we found
+            f(self, entry.path(), outdir, None)?
+        }
+
+        Ok(())
+    }
+
+    /// Determines the output path of a target based on consistent snapshot rules. Fails if the
+    /// target already exists in the repo with a different hash.
+    fn target_path(
+        &self,
+        input: &Path,
+        outdir: &Path,
+        target_filename: Option<&str>,
+    ) -> Result<PathBuf> {
+        let outdir = std::fs::canonicalize(outdir).context(error::AbsolutePath { path: outdir })?;
+
+        // If the caller requested a specific target filename, use that, otherwise use the filename
+        // component of the input path.
+        let file_name = if let Some(target_filename) = target_filename {
+            target_filename
+        } else {
+            input
                 .file_name()
+                .context(error::NoFileName { path: input })?
                 .to_str()
-                .context(error::PathUtf8 { path: entry.path() })?;
+                .context(error::PathUtf8 { path: input })?
+        };
 
-            // Use the file name to see if a target exists in the repo
-            // with that name. If so...
-            let repo_target = match repo_targets.get(file_name) {
-                Some(repo_target) => repo_target,
-                None => continue,
```

Comment:
I think I missed a change in behavior here.

Before, if the target's filename didn't exist in the repo, we'd `continue` here and not symlink.  With the new code, we of course skip the hash check, but we do proceed to call the given copy or link function.

I think this is a problem because calling `link_targets` / `copy_targets` on a directory shouldn't link/copy targets that weren't in the repo, or we're effectively enlarging the target directory for no reason because the extra files aren't *in* the repo.